net: dhcpv4: set syslog server for log_backend_net via dhcpv4 option

maass-hamburg opened the pull request:
This adds the option to use dhcpv4 option 7 (Log Server) to set the ip address of log_backend_net.

maass-hamburg committed:
log: net: add option to set server ip with struct sockaddr

Adds a function to set the syslog server ip with a struct sockaddr,
so that it can be set without a string.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
net: dhcpv4: set log backend net ip via option

Adds the option to set the log_backend_net ip via dhcpv4
option 7.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

rlubos commented on include/zephyr/logging/log_backend_net.h:
```diff
@@ -27,6 +28,20 @@ extern "C" {
 bool log_backend_net_set_addr(const char *addr);
 
+/**
+ * @brief Allows user to set a server IP address at runtime
```
I think that the brief description should differentiate the two functions,  something like
> Allows user to set a server IP address, provided as sockaddr structure, at runtime

And same for the original one
> Allows user to set a server IP address, provided as string, at runtime

rlubos commented on include/zephyr/logging/log_backend_net.h:
```diff
@@ -27,6 +28,20 @@ extern "C" {
+ *
+ * @param addr     Pointer to the sockaddr structure that contains the IP address.
+ *
+ * @return True if parsing could be done, false otherwise.
```
There's no parsing involved for this function.

rlubos commented on subsys/logging/backends/Kconfig.net:
```diff
@@ -49,6 +49,13 @@ config LOG_BACKEND_NET_AUTOSTART
 	  thread might block.
 
+config LOG_BACKEND_NET_USE_DHCPV4_OPTION
+	bool "Use DHCPv4 option 7 to configure syslog server"
```
"option 7" could be confusing w/o looking into RFC, please use option name ("Log Server Option").

rlubos commented on subsys/logging/backends/log_backend_net.c:
```diff
@@ -233,6 +242,28 @@ bool log_backend_net_set_addr(const char *addr)
+	} else if (addr->sa_family == AF_INET6) {
+		net_sin6(&server_addr)->sin6_port = htons(514);
+	} else {
+		LOG_ERR("Unknown address family");
```
That should actually be checked before we overwrite the server address.

rlubos commented on subsys/logging/backends/log_backend_net.c:
```diff
@@ -233,6 +242,28 @@ bool log_backend_net_set_addr(const char *addr)
+
+	memcpy(&server_addr, addr, sizeof(server_addr));
+
+	if (addr->sa_family == AF_INET) {
```
I think we should overwrite the port number provided in `addr` if it's unspecified (`0`).

rlubos commented on subsys/net/lib/dhcpv4/dhcpv4.c:
```diff
@@ -810,6 +818,45 @@ static bool dhcpv4_parse_options(struct net_pkt *pkt,
+				return false;
+			}
+
+			(void)memset(&log_server, 0, sizeof(log_server));
```
nit: could just initialize the log server with `{ 0 }` to avoid the memset.

rlubos commented on subsys/net/lib/dhcpv4/dhcpv4.c:
```diff
@@ -810,6 +818,45 @@ static bool dhcpv4_parse_options(struct net_pkt *pkt,
+			log_server.sin_family = AF_INET;
+			log_backend_net_set_ip((struct sockaddr *)&log_server);
+
+			const struct log_backend *backend = log_backend_net_get();
+
+			if (!log_backend_is_active(backend)) {
+
+				log_backend_init(backend);
+				log_backend_enable(backend, backend->cb->ctx, CONFIG_LOG_MAX_LEVEL);
+			}
```
I don't really think we should have this here - the option should only update the log server address, not decide whether to activate the network log backed or not.

maass-hamburg replied:
This will now only be executed if `CONFIG_LOG_BACKEND_NET_AUTOSTART` is enabled, because if `CONFIG_LOG_BACKEND_NET_SERVER` is not set (empty string) during the init the net logging backend will be deactivated.

rlubos replied:
Ok, it makes more sense now. At least the application which wants to have full control over the backend can disable the autostart feature.

jukkar commented on subsys/net/lib/dhcpv4/dhcpv4.c:
```diff
@@ -810,6 +818,45 @@ static bool dhcpv4_parse_options(struct net_pkt *pkt,
+
+			(void)memset(&log_server, 0, sizeof(log_server));
+
+			if (net_pkt_read(pkt, log_server.sin_addr.s4_addr, 4) ||
```
This is perfectly valid code but there is some Misra rule that says we should always check the value explicitly unless the value is boolean. So here it would be better to write
```
if (net_pkt_read(pkt, log_server.sin_addr.s4_addr, 4) < 0 ||
    net_pkt_skip(pkt, length - 4U) < 0) {
```
Currently there is no checker for these but we can be proactive to avoid future warnings.

jukkar commented on subsys/net/lib/dhcpv4/dhcpv4.c:
```diff
@@ -810,6 +818,45 @@ static bool dhcpv4_parse_options(struct net_pkt *pkt,
+			const struct log_backend *backend = log_backend_net_get();
+
+			if (!log_backend_is_active(backend)) {
+
```
extra empty line can be omitted

rlubos commented on subsys/net/lib/dhcpv4/dhcpv4.c:
```diff
@@ -44,6 +48,11 @@ static struct k_work_delayable timeout_work;
 
+#if defined(CONFIG_LOG_BACKEND_NET_USE_DHCPV4_OPTION) && \
+	defined(CONFIG_LOG_BACKEND_NET_AUTOSTART)
+extern const struct log_backend *log_backend_net_get(void);
```
That'd be the fourth place where this function is redeclared:
* https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/logging/backends/log_backend_net.c#L45
* https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/lib/config/init.c#L33
* https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/net/syslog_net/src/main.c#L23

I think it'd make sense to do a little cleanup at this point, move the declaration to `log_backend_net.h` instead, and remove the remaining ones as well (in a separate commit).

rlubos reviewed (commented):
One final request, otherwise looks ok.

maass-hamburg committed:
log: net: move log_backend_net_get() to header

This moves the declaration of log_backend_net_get() to
zephyr/logging/log_backend_net.h

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

jukkar commented on subsys/net/lib/dhcpv4/dhcpv4.c:
```diff
@@ -810,6 +814,43 @@ static bool dhcpv4_parse_options(struct net_pkt *pkt,
+			log_server.sin_family = AF_INET;
+			log_backend_net_set_ip((struct sockaddr *)&log_server);
+
+#ifdef CONFIG_LOG_BACKEND_NET_AUTOSTART
+			const struct log_backend *backend = log_backend_net_get();
+
+			if (!log_backend_is_active(backend)) {
+				log_backend_activate(backend, backend->cb->ctx);
+			}
+#endif /* CONFIG_LOG_BACKEND_NET_AUTOSTART */
```
This does not really belong to DHCP module. IMHO we should just call a function, could be called `log_backend_net_start()`, in net log backend, which then would have the these checks in them and do what is needed.

maass-hamburg committed:
log: net: Add log_backend_net_start function

Adds a function to start log_backend_net

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg commented:
rebase to resolve merge conflicts

maass-hamburg commented:
ping @nordic-krch
